File: AoC_2024/20/1.py
n = 141
INF = 1000000000

def Paths(v, cheat, maze, dist, bound, cnt):
    print(f"v: {v}")
    if bound == 0:
        return

    i = v // n
    j = v % n

    offsets = [(-1, 0), (0, -1), (1, 0), (0, 1)]
    for off_i, off_j in offsets:
        if 0 <= i + off_i < n and 0 <= j + off_j < n:
            u = (i + off_i) * n + (j + off_j)
            if maze[i][j] == '#':
                if cheat == 1 and dist[0][u] + 1 == dist[1][v] and dist[1][v] <= bound:
                    Paths(u, 0, maze, dist, bound - 1, cnt)

                if cheat == 2 and dist[1][u] + 1 == dist[2][v] and maze[i + off_i][j + off_j] == '#' and dist[2][v] <= bound:
                    cnt.add((v, u))

            else:
                if cheat == 0 and dist[0][u] + 1 == dist[0][v] and dist[1][v] <= bound:
                    Paths(u, 0, maze, dist, bound - 1, cnt)

                if cheat == 2 and dist[1][u] + 1 == dist[2][v] and maze[i + off_i][j + off_j] == '#' and dist[2][v] <= bound:
                    cnt.add((v, u))

                if cheat == 2 and dist[2][u] + 1 == dist[2][v] and dist[2][v] <= bound:
                    Paths(u, 2, maze, dist, bound - 1, cnt)

def Dijkstra(start, end, maze):
    dist = [[INF] * (n * n) for _ in range(3)]
    dist[0][start] = 0

    que = set()
    que.add((0, start, 0))

    while que:
        v = min(que, key=lambda x: x[0])[1]
        i = v // n
        j = v % n
        cheat = min(que, key=lambda x: x[0])[2]
        que.remove(min(que, key=lambda x: x[0]))
        offsets = [(-1, 0), (0, -1), (1, 0), (0, 1)]
        for off_i, off_j in offsets:
            if 0 <= i + off_i < n and 0 <= j + off_j < n:
                u = (i + off_i) * n + (j + off_j)
                if maze[i + off_i][j + off_j] == '#':
                    if dist[0][v] + 1 < dist[1][u]:
                        que.discard((dist[1][u], u, 1))
                        dist[1][u] = dist[0][v] + 1
                        que.add((dist[1][u], u, 1))

                    if dist[1][v] + 1 < dist[2][u] and maze[i][j] == '#':
                        que.discard((dist[2][u], u, 2))
                        dist[2][u] = dist[1][v] + 1
                        que.add((dist[2][u], u, 2))
                else:
                    if dist[0][v] + 1 < dist[0][u]:
                        que.discard((dist[0][u], u, 0))
                        dist[0][u] = dist[0][v] + 1
                        que.add((dist[0][u], u, 0))

                    if dist[1][v] + 1 < dist[2][u] and maze[i][j] == '#':
                        que.discard((dist[2][u], u, 2))
                        dist[2][u] = dist[1][v] + 1
                        que.add((dist[2][u], u, 2))

                    if dist[2][v] + 1 < dist[2][u]:
                        que.discard((dist[2][u], u, 2))
                        dist[2][u] = dist[2][v] + 1
                        que.add((dist[2][u], u, 2))

    ans = INF
    for i in range(3):
        ans = min(ans, dist[i][end])

    for i in range(n):
        for j in range(n):
            k = INF
            v = i * n + j
            for q in range(3):
                k = min(k, dist[q][v])
            if k == INF:
                print(maze[i][j], end='')
            elif dist[0][v] != INF:
                print('@', end='')
            else:
                print(maze[i][j], end='')
        print()

    bound = 84
    cnt = set()
    Paths(end, 1, maze, dist, bound, cnt)
    Paths(end, 2, maze, dist, bound, cnt)
    print(len(cnt))

    return ans

def Dijkstra_norm(start, end, maze):
    dist = [INF] * (n * n)
    dist[start] = 0

    que = set()
    que.add((0, start))

    while que:
        v = min(que, key=lambda x: x[0])[1]
        i = v // n
        j = v % n
        que.remove(min(que, key=lambda x: x[0]))
        offsets = [(-1, 0), (0, -1), (1, 0), (0, 1)]
        for off_i, off_j in offsets:
            if 0 <= i + off_i < n and 0 <= j + off_j < n:
                u = (i + off_i) * n + (j + off_j)
                if maze[i + off_i][j + off_j] != '#':
                    if dist[v] + 1 < dist[u]:
                        que.discard((dist[u], u))
                        dist[u] = dist[v] + 1
                        que.add((dist[u], u))

    ans = INF
    ans = min(ans, dist[end])

    return ans

def main():
    with open("test.txt", "r") as f:
        maze = [list(f.readline().strip()) for _ in range(n)]

    start = end = 0

    for i in range(n):
        for j in range(n):
            if maze[i][j] == 'E':
                end = i * n + j
            elif maze[i][j] == 'S':
                start = i * n + j

    print(start, end)

    cnt = 0
    mx_ans = Dijkstra_norm(start, end, maze)
    bound = mx_ans - 100
    for i in range(n):
        for j in range(n):
            if maze[i][j] != '#':
                continue
            offsets = [(-1, 0), (0, -1), (1, 0), (0, 1)]
            for off_i, off_j in offsets:
                if 0 <= i + off_i < n and 0 <= j + off_j < n:
                    save = maze[i + off_i][j + off_j]
                    maze[i][j] = '.'
                    maze[i + off_i][j + off_j] = '.'
                    if Dijkstra_norm(start, end, maze) <= bound:
                        cnt += 1
                    maze[i][j] = '#'
                    maze[i + off_i][j + off_j] = save

    print(cnt)

main()

File: AoC_2024/16/1.cpp
#include <iostream>
#include <vector>
#include <set>

const int n = 141;
const int INF = 1000000000;
int Dijkstra(int start, std::vector<int>& ends, std::vector<std::vector<std::pair<int, int>>>& g) {

  std::vector<int> dist(n * n * 4, INF);
  dist[start] = 0;

  std::set<std::pair<int, int>> que;
  que.insert({0, start});

  while (!que.empty()) {
    int v = que.begin() -> second;
    que.erase(que.begin());
    for (auto [w, u] : g[v]) {
      if (dist[v] + w < dist[u]) {
        que.erase({dist[u], u});
        dist[u] = dist[v] + w;
        que.insert({dist[u], u});
      }
    }
  }
  int ans = INF;
  for (int end : ends) {
    ans = std::min(ans, dist[end]);
  }
  return ans;
}


int main() {
  freopen("test.txt", "r", stdin);

  std::vector<std::vector<std::pair<int, int>>> g(n * n * 4);

  int start;
  std::vector<int> ends;
  std::vector<std::vector<char>> maze(n, std::vector<char>(n));
  for (int i = 0; i < n; ++i) {
    for (int j = 0; j < n; ++j) {
        std::cin >> maze[i][j];
        if (maze[i][j] == 'S') {
          start = (i * n + j) * 4 + 3;
        } else if (maze[i][j] == 'E') {
          for (int q = 0; q < 4; ++q) {
            ends.push_back((i * n + j) * 4 + q);
          }
        }
    }
  }

  for (int i = 0; i < n; ++i) {
    for (int j = 0; j < n; ++j) {
        if (maze[i][j] == '#') {
          continue;
        }
        int idx = (i * n + j) * 4;
        if (i != 0 && maze[i - 1][j] != '#') {
            int idx1 = ((i - 1) * n + j) * 4;
            g[idx].push_back({1, idx1 + 2});
        }
        if (i != n - 1 && maze[i + 1][j] != '#') {
            int idx1 = ((i + 1) * n + j) * 4;
            g[idx + 2].push_back({1, idx1});
        }
        if (j != 0 && maze[i][j - 1] != '#') {
            int idx1 = (i * n + j - 1) * 4;
            g[idx + 3].push_back({1, idx1 + 1});
        }
        if (j != n - 1 && maze[i][j + 1] != '#') {
            int idx1 = (i * n + j + 1) * 4;
            g[idx + 1].push_back({1, idx1 + 3});
        }
        g[idx].push_back({0, idx + 2});
        g[idx + 2].push_back({0, idx});
        g[idx + 1].push_back({0, idx + 3});
        g[idx + 3].push_back({0, idx + 1});

        g[idx].push_back({1000, idx + 1});
        g[idx].push_back({1000, idx + 3});

        g[idx + 1].push_back({1000, idx});
        g[idx + 1].push_back({1000, idx + 2});

        g[idx + 2].push_back({1000, idx + 1});
        g[idx + 2].push_back({1000, idx + 3});

        g[idx + 3].push_back({1000, idx});
        g[idx + 3].push_back({1000, idx + 2});
    }
  }
  std::cout << start << "\n";
  for (int i = 0; i < 4; ++i) {
      std::cout << ends[i] << " ";
  }

  std::cout << Dijkstra(start, ends, g);
}
